crates/pico_core/src/storage.rs

```diff
@@ -1,25 +1,160 @@
 use std::any::Any;
 
+use dashmap::DashMap;
+use once_map::OnceMap;
+
 use crate::{
-    container::Container,
     database::Database,
+    dependency_stack::DependencyStack,
     epoch::Epoch,
+    generation::Generation,
     key::Key,
-    node::{Dependency, DerivedNode, DerivedNodeId, SourceNode},
+    node::{Dependency, DerivedNode, DerivedNodeId, DerivedNodeRevision, SourceNode},
     params::ParamId,
 };
 
-pub trait Storage<Db: Database + ?Sized> {
-    fn derived_nodes(&self) -> &impl Container<DerivedNodeId, DerivedNode<Db>>;
-    fn source_nodes(&self) -> &impl Container<Key, SourceNode>;
-    fn params(&self) -> &impl Container<ParamId, Box<dyn Any>>;
-    fn current_epoch(&self) -> Epoch;
+#[derive(Debug)]
+pub struct Storage<Db: Database + ?Sized> {
+    pub current_epoch: Epoch,
+    pub dependency_stack: DependencyStack<(Epoch, Dependency)>,
+    pub params_map: DashMap<ParamId, Index>,
+    pub derived_nodes: DerivedNodesStore,
+    pub source_nodes: DashMap<Key, SourceNode>,
+    pub capacity: usize,
+    pub map: OnceMap<Epoch, Box<Generation<Db>>>,
+}
+
+impl<Db: Database + ?Sized> Storage<Db> {
+    pub fn new(capacity: usize) -> Self {
+        let map = OnceMap::new();
+        map.insert(Epoch::new(), |_| Box::new(Generation::new()));
+        Self {
+            current_epoch: Epoch::new(),
+            dependency_stack: DependencyStack::new(),
+            params_map: DashMap::default(),
+            derived_nodes: DerivedNodesStore::default(),
+            source_nodes: DashMap::default(),
+            capacity,
+            map,
+        }
+    }
+}
+
+impl<Db: Database + ?Sized> Storage<Db> {
+    pub fn increment_epoch(&mut self) -> Epoch {
+        let current_epoch = self.current_epoch.increment();
+        self.map
+            .insert(current_epoch, |_| Box::new(Generation::new()));
+        // TODO: not the best way to do it, worth to create something like OnceOrderedMap
+        if self.map.read_only_view().len() >= self.capacity {
```

Comment:
IMO it's unexpected that `increment_epoch` would have such consequences...

Reply:
Agreed. As we discussed, it is better to have a GC task to preserve more information during user's typing. Need to think about a good API to expose.